### PR TITLE
Fixed #24064 - Prevented database access in compile time in spatialite models

### PR DESCRIPTION
**Pull request by coldmind**
Here is the patch that can fix problem.
Ideas about improvements are welcome.

### CODE REVIEW

**Comment by coldmind** [comment]
@charettes, well, it works, but I don't understand the reason to check `if connection_created.disconnect` inside the receiver function.

**Comment by coldmind** [comment]
buildbot, test this please

**Comment by coldmind** [comment]
buildbot, test this please.

**Comment by timgraham** [comment]
buildbot, test on trusty.

**Comment by charettes** [comment]
@coldmind, it's to prevent two concurrent threads from creating a race condition where the signal is dispatched twice.

**Comment by timgraham** [comment]
buildbot, test on trusty.

**Comment by coldmind** [comment]
@charettes, @timgraham, due to tests, disconnection is called once, so seems no need to have check if field already added to model (see diff above). But I'm afraid there can be situation where it could be. What do you think?

**Comment by timgraham** [comment]
Simon, is this a different problem from what [dispatch_uid](https://docs.djangoproject.com/en/dev/topics/signals/#preventing-duplicate-signals) attempts to solve?

**Comment by charettes** [comment]
>  due to tests, disconnection is called once, so seems no need to have check if field already added to model (see diff above). But I'm afraid there can be situation where it could be. What do you think?

> Simon, is this a different problem from what dispatch_uid attempts to solve?

Yes, `dispatch_uid` is used to prevent multiple registration of the same receiver to a signal. This was handy when your project could be accessible from multiple locations on your `PYTHONPATH`. It's a feature similar to the one use in `ModelBase.__new__` to prevent multiple registrations of the same model given the module containing them is imported multiple times due to `PYTHONPATH` misconfiguration.

The is trying to prevent a race condition in a multi-threaded environment (not reproducible during tests) such as `runserver` where `n` connections are created and dispatch the `connection_created` signal twice before it's disconnected. In this case you would end up with `n` times the fields you wanted to add once.

**Comment by charettes** [comment]
buildbot, test on trusty.

**Comment by charettes** [comment]
I re-retriggered the build since the failures seems unrelated.

**Comment by timgraham** [comment]
Could you move the `django/dispatch/dispatcher.py` changes to a separate commit that precedes the other and add some tests for it? Otherwise, I think this is good to go.

**Comment by coldmind** [comment]
@timgraham, sure, going to add some tests for dispatcher.

**Comment by coldmind** [comment]
buildbot, test on trusty.

**Comment by coldmind** [comment]
buildbot, test this please.

**Comment by claudep** [comment]
Note that disconnect is documented https://docs.djangoproject.com/en/dev/topics/signals/#disconnecting-signals, an update is needed there too.

File: django/dispatch/dispatcher.py [hunk 1]
```diff
@@ -165,9 +165,13 @@ def disconnect(self, receiver=None, sender=None, weak=True, dispatch_uid=None):
             for index in range(len(self.receivers)):
                 (r_key, _) = self.receivers[index]
                 if r_key == lookup_key:
+                    disconnected = True
                     del self.receivers[index]
                     break
+            else:
+                disconnected = False
```
Comment (apollo13) [inline django/dispatch/dispatcher.py]: You could move this line above the for loop and get rid of the else that way.

**Comment by coldmind** [comment]
buildbot, test this please.

**Comment by coldmind** [comment]
jenkins failed again with `stderr: fatal: reference is not a tree`

**Comment by timgraham** [comment]
No worries, this is likely fine. Will commit shortly.

**Comment by timgraham** [comment]
Merged in 23f1a8dad223ee79cd206c9032c871c874f7e7af and 839f431ef5b927c8b07ce33483b3a7b0dd17b761. Thanks very much for investigating the issue!

**Pull request closed without merging**